src/fromager/resolver.py

```diff
@@ -509,3 +509,52 @@ def _find_tags(
                 yield entry["tarball_url"], version
             # pagination links
             nexturl = resp.links.get("next", {}).get("url")
+
+
+class CustomPackageProvider(BaseProvider):
+    """Get best match"""
+
+    def __init__(
+        self,
+        req: Requirement,
+        constraints: Constraints,
+        versions: dict[Version, str],
```

Comment:
We have #489 describing creating a class that maps versions to arbitrary values and offering a way to resolve to the "right" value. That would invert the direction of this PR, since the dict-like object would be the main interface. I wonder if that would provide a more natural API for consumers for this use case. We could hide a GenericProvider inside of the VersionMap type, for example.

Reply:
I added VersionsMap to see if it made a big difference -- it doesn't AFAICT so I've submitted a v2 with that change.

Reply:
I'll leave it up to you to resolve this comment @dhellmann -- I'm not sure if I captured your concern or not.